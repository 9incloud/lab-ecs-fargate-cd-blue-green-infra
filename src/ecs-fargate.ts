import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecr from '@aws-cdk/aws-ecr';
import * as ecs from '@aws-cdk/aws-ecs';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';
import { NetworkConfig, StageConfig } from './main';

interface EcsFargateProps extends cdk.StackProps {
  stageConfig: StageConfig;
  readonly vpc?: ec2.Vpc;
  readonly alb?: elbv2.IApplicationLoadBalancer;
}

export class EcsFargate extends cdk.Stack {
  readonly service: ecs.FargateService;
  readonly ecrRepository: ecr.Repository;

  constructor(scope: cdk.Construct, id: string, props: EcsFargateProps) {
    super(scope, id, props);

    this.ecrRepository = new ecr.Repository(this, 'Repository', {
      repositoryName: props.stageConfig.Ecs.ecrRepositoryName,
    });

    const cluster = new ecs.Cluster(this, 'Cluster', {
      vpc:
        props?.vpc ??
        ec2.Vpc.fromLookup(this, 'DefaultVpc', { isDefault: true }),
    });

    const fargateTaskDefinition = new ecs.FargateTaskDefinition(
      this,
      'TaskDef',
      {
        memoryLimitMiB: props.stageConfig.Ecs.memoryLimitMiB,
        cpu: props.stageConfig.Ecs.cpu,
      }
    );

    fargateTaskDefinition.addContainer(props.stageConfig.Ecs.container.name, {
      image: ecs.ContainerImage.fromEcrRepository(this.ecrRepository),
      environment: props.stageConfig.Ecs.container.environment,
      portMappings: [{ containerPort: 80, hostPort: 80 }],
      logging: ecs.LogDriver.awsLogs({
        logGroup: new logs.LogGroup(this, 'EcsLogGroup', {
          logGroupName: props.stageConfig.Ecs.container.logGroupName,
        }),
        streamPrefix: 'ecs',
      }),
    });

    this.service = new ecs.FargateService(this, 'Service', {
      cluster,
      taskDefinition: fargateTaskDefinition,
      desiredCount: props.stageConfig.Ecs.service.desiredCount,
      circuitBreaker: {
        rollback: props.stageConfig.Ecs.service.circuitBreakerRollback,
      },
      minHealthyPercent: props.stageConfig.Ecs.service.minHealthyPercent,
      maxHealthyPercent: props.stageConfig.Ecs.service.maxHealthyPercent,
      healthCheckGracePeriod: cdk.Duration.seconds(0),
      assignPublicIp: true,
    });

    if (props?.alb) {
      this.createAlbListenerAndTargetGroup(
        props.alb,
        this.service,
        props.stageConfig.Network
      );
    }
  }

  private createAlbListenerAndTargetGroup(
    alb: elbv2.IApplicationLoadBalancer,
    service: ecs.FargateService,
    networkProps: NetworkConfig
  ): void {
    const listener = new elbv2.ApplicationListener(this, 'AlbListener', {
      loadBalancer: alb,
      port: 443,
      sslPolicy: elbv2.SslPolicy.RECOMMENDED,
    });

    if (networkProps.alb?.listener?.certificateArn) {
      listener.addCertificates('AlbListenerCertification', [
        {
          certificateArn: networkProps.alb.listener.certificateArn,
        },
      ]);
    }

    new elbv2.ApplicationListenerRule(this, 'AlbListenerRule', {
      listener: listener,
      priority: 10,
      action: elbv2.ListenerAction.fixedResponse(200, {
        contentType: 'application/json',
        messageBody: '',
      }),
      conditions: [elbv2.ListenerCondition.pathPatterns(['/status'])],
    });

    listener.addTargets('ECS', {
      port: 80,
      targets: [service],
      healthCheck: {
        enabled: networkProps.alb.targetHealthCheck.enabled,
        interval: cdk.Duration.seconds(
          networkProps.alb.targetHealthCheck.interval
        ),
        path: networkProps.alb.targetHealthCheck.path,
        timeout: cdk.Duration.seconds(
          networkProps.alb.targetHealthCheck.timeout
        ),
        healthyThresholdCount:
          networkProps.alb.targetHealthCheck.healthyThresholdCount,
        unhealthyThresholdCount:
          networkProps.alb.targetHealthCheck.unhealthyThresholdCount,
        healthyHttpCodes: '200',
      },
      deregistrationDelay: cdk.Duration.seconds(
        networkProps.alb.deregistrationDelay
      ),
    });
  }
}
