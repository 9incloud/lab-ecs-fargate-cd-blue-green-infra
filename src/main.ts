import { App } from '@aws-cdk/core';
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import { EcsFargate } from './ecs-fargate';
import { RestAPINetwork } from './restapi-network';

enum Stage {
  STAGING = 'staging',
  PROD = 'prod',
}

export interface StageConfig {
  Network: NetworkConfig;
  Ecs: EcsConfig;
}

export interface NetworkConfig {
  vpc: {
    maxAzs: number;
    natGateways: number;
    ipv6enabled: boolean;
  };
  alb: {
    listener: {
      certificateArn: string;
    };
    targetHealthCheck: {
      enabled: boolean;
      interval: number;
      path: string;
      timeout: number;
      healthyThresholdCount: number;
      unhealthyThresholdCount: number;
    };
  };
}

interface EcsConfig {
  ecrRepositoryName: string;
  memoryLimitMiB: number;
  cpu: number;
  executionRoleArn: string;
  taskRole: {
    customManagedPolicies: string[];
  };
  container: {
    environment: {
      [key: string]: string;
    };
    logGroupName: string;
  };
  service: {
    desiredCount: number;
    minHealthyPercent: number;
    maxHealthyPercent: number;
    circuitBreakerRollback: boolean;
  };
}

const stage = process.env.STAGE || null;

if (stage === null || !Object.values(Stage).includes(stage as Stage)) {
  throw new Error('STAGE environment variable is required');
}

const stageConfig: StageConfig = readConfig(stage) as StageConfig;

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

const restAPINetwork = new RestAPINetwork(app, 'api-network', {
  stageConfig,
  env: devEnv,
});

const ecsFargate = new EcsFargate(app, 'api-app', {
  stageConfig,
  vpc: restAPINetwork.vpc,
  alb: restAPINetwork.alb,
  env: devEnv,
});

ecsFargate.addDependency(restAPINetwork);

app.synth();

function readConfig(stageName: string): any {
  return yaml.safeLoad(
    fs.readFileSync(`${__dirname}/../configs/${stageName}.yml`, 'utf8')
  );
}
