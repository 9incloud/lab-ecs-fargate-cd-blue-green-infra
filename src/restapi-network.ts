import * as ec2 from '@aws-cdk/aws-ec2';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as cdk from '@aws-cdk/core';
import { StageConfig } from './main';

interface RestAPINetworkProps extends cdk.StackProps {
  stageConfig: StageConfig;
}

export class RestAPINetwork extends cdk.Stack {
  vpc: ec2.Vpc;
  alb: elbv2.ApplicationLoadBalancer;

  constructor(scope: cdk.Construct, id: string, props: RestAPINetworkProps) {
    super(scope, id, props);
    this.vpc = new ec2.Vpc(this, 'Vpc', {
      cidr: '10.0.0.0/16',
      maxAzs: props.stageConfig.Network.vpc.maxAzs,
      natGateways: props.stageConfig.Network.vpc.natGateways,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'ingress',
          subnetType: ec2.SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'application',
          subnetType: ec2.SubnetType.PRIVATE,
        },
        {
          cidrMask: 28,
          name: 'rds',
          subnetType: ec2.SubnetType.ISOLATED,
        },
      ],
    });

    if (props.stageConfig.Network.vpc.ipv6enabled) this.enableIpv6(this.vpc);

    this.alb = new elbv2.ApplicationLoadBalancer(this, 'Alb', {
      vpc: this.vpc,
      internetFacing: true,
      ipAddressType: elbv2.IpAddressType.DUAL_STACK,
      vpcSubnets: this.vpc.selectSubnets({
        subnetType: ec2.SubnetType.PUBLIC,
      }),
    });
  }

  enableIpv6(vpc: ec2.Vpc) {
    const cfnIpv6Cidr = new ec2.CfnVPCCidrBlock(this, 'Ipv6Cidr', {
      vpcId: vpc.vpcId,
      amazonProvidedIpv6CidrBlock: true,
    });

    this.vpc.publicSubnets.forEach((subnet, idx) => {
      const vpcCidrBlock = cdk.Fn.select(0, this.vpc.vpcIpv6CidrBlocks);
      const ipv6Cidrs = cdk.Fn.cidr(
        vpcCidrBlock,
        this.vpc.publicSubnets.length,
        '64'
      );
      const cfnSubnet = subnet.node.defaultChild as ec2.CfnSubnet;
      cfnSubnet.ipv6CidrBlock = cdk.Fn.select(idx, ipv6Cidrs);
      cfnSubnet.addDependsOn(cfnIpv6Cidr);
    });

    cfnIpv6Cidr.addDependsOn(this.vpc.node.defaultChild as ec2.CfnVPC);
  }
}
