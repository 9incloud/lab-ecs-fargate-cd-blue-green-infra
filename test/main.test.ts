import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import { Pipeline } from '../src//pipeline';
import { EcsFargate } from '../src/ecs-fargate';
import { StageConfig } from '../src/main';
import {} from '../src/pipeline';
import { RestAPINetwork } from '../src/restapi-network';

test('Snapshot', () => {
  process.env.STAGE = 'lab';

  const stageConfig: StageConfig = yaml.safeLoad(
    fs.readFileSync(`${__dirname}/../configs/${process.env.STAGE}.yml`, 'utf8')
  ) as StageConfig;

  const env = {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  };

  const app = new App();
  const stackNetwork = new RestAPINetwork(app, 'testStack1', {
    env,
    stageConfig,
  });

  const stackFargate = new EcsFargate(app, 'testStack2', {
    vpc: stackNetwork.vpc,
    alb: stackNetwork.alb,
    env,
    stageConfig,
  });

  const pipeline = new Pipeline(app, 'ApiPipeline', {
    fargateService: stackFargate.service,
    ecrRepository: stackFargate.ecrRepository,
    env,
  });

  expect(SynthUtils.toCloudFormation(stackNetwork)).toMatchSnapshot();
  expect(SynthUtils.toCloudFormation(stackFargate)).toMatchSnapshot();
  expect(SynthUtils.toCloudFormation(pipeline)).toMatchSnapshot();
});
