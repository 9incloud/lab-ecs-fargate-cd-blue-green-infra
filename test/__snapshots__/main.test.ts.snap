// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Outputs": Object {
    "ExportsOutputFnGetAttalbSecurityGroup49866104GroupIdD5B19856": Object {
      "Export": Object {
        "Name": "testStack1:ExportsOutputFnGetAttalbSecurityGroup49866104GroupIdD5B19856",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "albSecurityGroup49866104",
          "GroupId",
        ],
      },
    },
    "ExportsOutputRefalb8A8B13C2253BCEDB": Object {
      "Export": Object {
        "Name": "testStack1:ExportsOutputRefalb8A8B13C2253BCEDB",
      },
      "Value": Object {
        "Ref": "alb8A8B13C2",
      },
    },
    "ExportsOutputRefvpcA2121C384D1B3CDE": Object {
      "Export": Object {
        "Name": "testStack1:ExportsOutputRefvpcA2121C384D1B3CDE",
      },
      "Value": Object {
        "Ref": "vpcA2121C38",
      },
    },
    "ExportsOutputRefvpcingressSubnet1Subnet4E0162281DAD06BF": Object {
      "Export": Object {
        "Name": "testStack1:ExportsOutputRefvpcingressSubnet1Subnet4E0162281DAD06BF",
      },
      "Value": Object {
        "Ref": "vpcingressSubnet1Subnet4E016228",
      },
    },
    "ExportsOutputRefvpcingressSubnet2SubnetEC320D0340041730": Object {
      "Export": Object {
        "Name": "testStack1:ExportsOutputRefvpcingressSubnet2SubnetEC320D0340041730",
      },
      "Value": Object {
        "Ref": "vpcingressSubnet2SubnetEC320D03",
      },
    },
  },
  "Resources": Object {
    "Ipv6Cidr": Object {
      "DependsOn": Array [
        "vpcA2121C38",
      ],
      "Properties": Object {
        "AmazonProvidedIpv6CidrBlock": true,
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::VPCCidrBlock",
    },
    "alb8A8B13C2": Object {
      "DependsOn": Array [
        "vpcingressSubnet1DefaultRoute69B9019B",
        "vpcingressSubnet2DefaultRoute9BA4F3D4",
      ],
      "Properties": Object {
        "IpAddressType": "dualstack",
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "albSecurityGroup49866104",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          Object {
            "Ref": "vpcingressSubnet1Subnet4E016228",
          },
          Object {
            "Ref": "vpcingressSubnet2SubnetEC320D03",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "albSecurityGroup49866104": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB testStack1alb037B64BD",
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
          Object {
            "CidrIpv6": "::/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "vpcA2121C38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "testStack1/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "vpcIGWE57CBDCA": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "testStack1/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "vpcVPCGW7984C166": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "vpcingressSubnet1DefaultRoute69B9019B": Object {
      "DependsOn": Array [
        "vpcVPCGW7984C166",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": Object {
          "Ref": "vpcingressSubnet1RouteTable1A2E6413",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcingressSubnet1RouteTable1A2E6413": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "testStack1/vpc/ingressSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcingressSubnet1RouteTableAssociationBA5D4A44": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcingressSubnet1RouteTable1A2E6413",
        },
        "SubnetId": Object {
          "Ref": "vpcingressSubnet1Subnet4E016228",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcingressSubnet1Subnet4E016228": Object {
      "DependsOn": Array [
        "Ipv6Cidr",
      ],
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/24",
        "Ipv6CidrBlock": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::Cidr": Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::GetAtt": Array [
                        "vpcA2121C38",
                        "Ipv6CidrBlocks",
                      ],
                    },
                  ],
                },
                2,
                "64",
              ],
            },
          ],
        },
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "testStack1/vpc/ingressSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcingressSubnet2DefaultRoute9BA4F3D4": Object {
      "DependsOn": Array [
        "vpcVPCGW7984C166",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": Object {
          "Ref": "vpcingressSubnet2RouteTable4047C3AC",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcingressSubnet2RouteTable4047C3AC": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "testStack1/vpc/ingressSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcingressSubnet2RouteTableAssociationBD7DE3FC": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcingressSubnet2RouteTable4047C3AC",
        },
        "SubnetId": Object {
          "Ref": "vpcingressSubnet2SubnetEC320D03",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcingressSubnet2SubnetEC320D03": Object {
      "DependsOn": Array [
        "Ipv6Cidr",
      ],
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/24",
        "Ipv6CidrBlock": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::Cidr": Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::GetAtt": Array [
                        "vpcA2121C38",
                        "Ipv6CidrBlocks",
                      ],
                    },
                  ],
                },
                2,
                "64",
              ],
            },
          ],
        },
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "testStack1/vpc/ingressSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;

exports[`Snapshot 2`] = `
Object {
  "Resources": Object {
    "AlbListenerDFCDF14C": Object {
      "Properties": Object {
        "Certificates": Array [
          Object {
            "CertificateArn": "arn:aws:acm:ap-northeast-1:482631629698:certificate/6ba7f53f-b622-4627-b880-955c382245de",
          },
        ],
        "DefaultActions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "AlbListenerECSGroup21012B54",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": Object {
          "Fn::ImportValue": "testStack1:ExportsOutputRefalb8A8B13C2253BCEDB",
        },
        "Port": 443,
        "Protocol": "HTTPS",
        "SslPolicy": "ELBSecurityPolicy-2016-08",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "AlbListenerECSGroup21012B54": Object {
      "Properties": Object {
        "HealthCheckEnabled": true,
        "HealthCheckIntervalSeconds": 30,
        "HealthCheckPath": "/hello",
        "HealthCheckTimeoutSeconds": 5,
        "HealthyThresholdCount": 5,
        "Matcher": Object {
          "HttpCode": "200",
        },
        "Port": 80,
        "Protocol": "HTTP",
        "TargetType": "ip",
        "UnhealthyThresholdCount": 2,
        "VpcId": Object {
          "Fn::ImportValue": "testStack1:ExportsOutputRefvpcA2121C384D1B3CDE",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "AlbListenerRuleE27FB4D8": Object {
      "Properties": Object {
        "Actions": Array [
          Object {
            "FixedResponseConfig": Object {
              "ContentType": "application/json",
              "MessageBody": "",
              "StatusCode": "200",
            },
            "Type": "fixed-response",
          },
        ],
        "Conditions": Array [
          Object {
            "Field": "path-pattern",
            "PathPatternConfig": Object {
              "Values": Array [
                "/status",
              ],
            },
          },
        ],
        "ListenerArn": Object {
          "Ref": "AlbListenerDFCDF14C",
        },
        "Priority": 10,
      },
      "Type": "AWS::ElasticLoadBalancingV2::ListenerRule",
    },
    "ClusterEB0386A7": Object {
      "Type": "AWS::ECS::Cluster",
    },
    "EcsLogGroupD0138CE7": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": "/ecs/staging-wondercise-api",
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "Repository22E53BBD": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RepositoryName": "wondercise-rest-api",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
    "ServiceD69D759B": Object {
      "DependsOn": Array [
        "AlbListenerECSGroup21012B54",
        "AlbListenerDFCDF14C",
      ],
      "Properties": Object {
        "Cluster": Object {
          "Ref": "ClusterEB0386A7",
        },
        "DeploymentConfiguration": Object {
          "DeploymentCircuitBreaker": Object {
            "Enable": true,
            "Rollback": true,
          },
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 100,
        },
        "DeploymentController": Object {
          "Type": "ECS",
        },
        "DesiredCount": 0,
        "EnableECSManagedTags": false,
        "HealthCheckGracePeriodSeconds": 0,
        "LaunchType": "FARGATE",
        "LoadBalancers": Array [
          Object {
            "ContainerName": "restApiContainer",
            "ContainerPort": 80,
            "TargetGroupArn": Object {
              "Ref": "AlbListenerECSGroup21012B54",
            },
          },
        ],
        "NetworkConfiguration": Object {
          "AwsvpcConfiguration": Object {
            "AssignPublicIp": "DISABLED",
            "SecurityGroups": Array [
              Object {
                "Fn::GetAtt": Array [
                  "ServiceSecurityGroupC96ED6A7",
                  "GroupId",
                ],
              },
            ],
            "Subnets": Array [
              Object {
                "Fn::ImportValue": "testStack1:ExportsOutputRefvpcingressSubnet1Subnet4E0162281DAD06BF",
              },
              Object {
                "Fn::ImportValue": "testStack1:ExportsOutputRefvpcingressSubnet2SubnetEC320D0340041730",
              },
            ],
          },
        },
        "TaskDefinition": Object {
          "Ref": "TaskDef54694570",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "ServiceSecurityGroupC96ED6A7": Object {
      "Properties": Object {
        "GroupDescription": "testStack2/Service/SecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "testStack1:ExportsOutputRefvpcA2121C384D1B3CDE",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ServiceSecurityGroupfromtestStack1albSecurityGroup6E2410E3800E6C0DAC": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "FromPort": 80,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "ServiceSecurityGroupC96ED6A7",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::ImportValue": "testStack1:ExportsOutputFnGetAttalbSecurityGroup49866104GroupIdD5B19856",
        },
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "ServiceSecurityGrouptestStack1albSecurityGroup6E2410E380fromFD570F91": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "ServiceSecurityGroupC96ED6A7",
            "GroupId",
          ],
        },
        "FromPort": 80,
        "GroupId": Object {
          "Fn::ImportValue": "testStack1:ExportsOutputFnGetAttalbSecurityGroup49866104GroupIdD5B19856",
        },
        "IpProtocol": "tcp",
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "TaskDef54694570": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "ASPNETCORE_ENVIRONMENT",
                "Value": "Staging",
              },
              Object {
                "Name": "COMPlus_ThreadPool_ForceMinWorkerThreads",
                "Value": "600",
              },
            ],
            "Essential": true,
            "Image": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Fn::Select": Array [
                      4,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "Repository22E53BBD",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".dkr.ecr.",
                  Object {
                    "Fn::Select": Array [
                      3,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "Repository22E53BBD",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".",
                  Object {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  Object {
                    "Ref": "Repository22E53BBD",
                  },
                  ":latest",
                ],
              ],
            },
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-group": Object {
                  "Ref": "EcsLogGroupD0138CE7",
                },
                "awslogs-region": Object {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "ecs",
              },
            },
            "Name": "restApiContainer",
            "PortMappings": Array [
              Object {
                "ContainerPort": 80,
                "HostPort": 80,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "Cpu": "256",
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefExecutionRoleB4775C97",
            "Arn",
          ],
        },
        "Family": "testStack2TaskDefE1A7FF4E",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": Array [
          "FARGATE",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefTaskRole1EDB4A67",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefExecutionRoleB4775C97": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskDefExecutionRoleDefaultPolicy0DBB737A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "Repository22E53BBD",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "EcsLogGroupD0138CE7",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskDefExecutionRoleDefaultPolicy0DBB737A",
        "Roles": Array [
          Object {
            "Ref": "TaskDefExecutionRoleB4775C97",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskDefTaskRole1EDB4A67": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
